apollo-router-core/src/cache.rs

```diff
@@ -0,0 +1,174 @@
+    ///
+    /// resolver is used to resolve cache misses.
+    /// cache_limit specifies the size (number of items) of the cache
+    pub fn new(resolver: Box<(dyn CacheResolver<K, V> + Send + Sync)>, cache_limit: usize) -> Self {
+        Self {
+            cached: Mutex::new(LruCache::new(cache_limit)),
+            wait_map: Mutex::new(HashMap::new()),
+            cache_limit,
+            resolver,
+        }
+    }
+
+    /// Get a value from the cache.
+    pub async fn get(&self, key: K) -> CacheResult<V> {
+        let mut locked_cache = self.cached.lock().await;
+        if let Some(value) = locked_cache.get(&key).cloned() {
+            return value;
+        }
+
+        // Holding a lock across the delegated get is a bad idea because
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        // We must only drop the locked cache after we have locked the
+        // wait map. Otherwise,we might get a race that causes us to
+        // miss a broadcast.
+        drop(locked_cache);
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                // Our use case is very specific, so we are sure
+                // that we won't get any errors here.
+                let (recv_key, recv_plan) = receiver.recv().await.expect(
+                    "the sender won't ever be dropped before all the receivers finish; qed",
+                );
+                debug_assert_eq!(recv_key, key);
+                recv_plan
+            }
+            None => {
+                let (tx, _rx) = broadcast::channel(1);
+                locked_wait_map.insert(key.clone(), tx.clone());
+                drop(locked_wait_map);
+                // This is the potentially high duration operation where we ask our resolver to
+                // resolve the key (retrieve a value) for us
+                // No cache locks are held here
+                let value = self.resolver.retrieve(key.clone()).await;
+                // Update our cache
```

Comment:
could this be done in a different scope? Like this:

```rust
{
                let mut locked_cache = self.cached.lock().await;
                locked_cache.put(key.clone(), value.clone());
                // Update our wait list
                let mut locked_wait_map = self.wait_map.lock().await;
                locked_wait_map.remove(&key);
}
```

Because with the current code, the cache and map locks are still held when we're await'ing on the blocking task

Reply:
I'm concerned that we would introduce a race if we didn't keep both the locks until the end of the notification. There may be opportunities for optimisation here, but I don't want to take the risk since it would only be a micro-optimisation and require careful analysis to make sure there were no mistakes.

Reply:
I don't think there will be a race here:
- the sender was removed from the map after the value was put in the cache, so any new queries for the same key will get it from the cache directly
- all the receivers have already dropped the locks when they're waiting

Reply:
I don't think so either, but I'm not sure. So, rather than take the risk of being wrong, I don't think the benefit of a small improvement in scalability is justified for the risk.

In particular, I'm concerned that a cache expiration event would cause a race. Very unlikely, but requires some careful thinking that I haven't done.

Reply:
yes the concerns are real. Actually, could we work on a proof that it works, even without the modification I want? Not necessarily something formal, just exploring the various cases and check that they will be safe

Reply:
ok so I tried to explore the various states. That looks like the problem where we should throw some formal method tool like TLA+ if we want to look at cases with more than 2 queries.

Here's the list of operations:
- Cl: lock the cache
- Cu: unlock the cache
- Wl: lock the wait map
- Wu: unlock the wait map
- Ch: cache hit
- Cm: cache miss
- Wh: wait map hit
- Wm: Wait map miss
- Ci: cache insert
- Wi: wait map insert
- Wr: wait map remove
- S: subscribe to the broadcast
- Rx: receive
- Tx: broadcast to all the receivers
- G: get from the resolver


There should be three flows:
- cache hit: ClChCu then done
- cache miss, first query to ask the cache, set up the broadcast, do the work, send values:
ClCmWlCu WmWiWu G ClCiWlWrTxWuCu
- cache miss, another query is currently working:
ClCmWlCu WhSWu Rx

Now I'll try variations with concurrent queries. First thing we can consider: the
sequence ClCmWlCu is completely protected by the cache lock, we cannot see multiple
queries inside at the same time. When a query reaches the Rx state (waiting for another
query to do the work), we can consider it safe, it won't lock anything else

1ClCmWlCu 2ClCm 2Wl: can't happen, 1 has Wl

1ClCmWlCu 2ClCm 1WmWiWuG 1Cl: can't happen, 2 has Cl
1ClCmWlCu 2ClCm 1WmWiWuG 2WlCu 2Wm: can't happen, 1 has Wi

1ClCmWlCu 2ClCm 1WmWiWuG 2WlCuWhSWu: from here, no issue, 2 is not waiting on any locks

1ClCmWlCu 2ClCm 1WmWiWuG 2WlCu 2WhS 1ClCi 1Wl: can't happen, 2 has Wl
1ClCmWlCu 2ClCm 1WmWiWuG 2WlCu 2WhSWuRx 1ClCiWlWrTxWuCu OK 

1ClCmWlCu 1WmWiWuG 2ClCmWl 1Cl: can't happen, 2 has Cl
1ClCmWlCu 1WmWiWuG 2ClCmWlCu 1ClCi 1Wl: can't happen, 2 has Wl

1ClCmWlCu 1WmWiWuG 2ClCmWlCu 1ClCi 2WhSWuRx 1WlWrTxWuCu OK

Reply:
the change I propose would affect the last sequence for the first query: ClCiWlWrTxWuCu would become ClCiWlWrWuCuTx

In all the cases I tried, once 1 reaches Wr, 2 is already in a safe state

Reply:
another note: the steps 2ClCm and 1WmWiWuG are commutative.

and actually the 3 queries case resolves to less cases, since it's very constrained:
1ClCmWlCu 2ClCm 1WmWiWuG 2WlCu 3ClCm 1Cl: can't happen 3 has Cl
1ClCmWlCu 2ClCm 1WmWiWuG 2WlCu 3ClCm 3Wl: can't happen 2 has Wl
1ClCmWlCu 2ClCm 1WmWiWuG 2WlCu 3ClCm 2WhSWuRx 3WlCu 1ClCi 1Wl: can't happen 3 has Wl
1ClCmWlCu 2ClCm 1WmWiWuG 2WlCu 3ClCm 2WhSWuRx 3WlCu 1ClCi 3WhSWuRx 1WlWrTxWuCu

Reply:
I'm having a hard time to follow the discussion, is there a chance we can land the "conservative approach" and chase the optimisation as a followup?

The code as is looks like a great improvement already IMO

Reply:
it is not just an optimization, but a way to avoid high contention on the cache lock: the broadcast channel works by cloning the value before sending it to each receiver, so if the value is large and there is a lot of receivers, the lock would be held while executing a serie of clones, which is potentially expensive. And so no other queries would be able to use the cache in that time

Reply:
This discussion has been extracted and deferred into issue: #244